hazelcast-sql/src/main/java/com/hazelcast/sql/impl/type/QueryDataType.java

```diff
@@ -100,12 +116,20 @@ public class QueryDataType implements IdentifiedDataSerializable, Serializable {
     public static final QueryDataType JSON = new QueryDataType(JsonConverter.INSTANCE);
     public static final QueryDataType ROW = new QueryDataType(RowConverter.INSTANCE);
 
+    private static final Map<String, QueryDataType> TYPES_BY_NAME = getEnumConstants(QueryDataType.class);
+    private static final Map<QueryDataType, String> NAMES =
+            TYPES_BY_NAME.entrySet().stream().collect(toMap(Entry::getValue, Entry::getKey));
+    private static final QueryDataType[] TYPES = TYPES_BY_NAME.values().stream()
+            .sorted(comparingInt(type -> type.converter.getId())).toArray(QueryDataType[]::new);
+
     private Converter converter;
-    // nonnull for custom types (nested types)
+    /** Nonnull for custom types. */
     private String objectTypeName;
     private TypeKind objectTypeKind = TypeKind.NONE;
     private String objectTypeMetadata;
-    private final List<QueryDataTypeField> objectFields = new ArrayList<>();
+    private List<QueryDataTypeField> objectFields;
```

Comment:
The current design using NonTraversableList and the finalize function seems a bit confusing. As I understand it, the goal is to allow adding data before finalization but prevent data traversal, and after finalization, prevent adding data but allow traversal. 

I'd like to propose more straightforward alternatives:
Use an Immutable List builder, where adding a field involves adding an object to the builder, and finalizing is calling the builder to create the list.
Create a function, like addAll(List<*> list), which can be called once to add all fields at once in an immutable list. Ensure that the function is only called when the field is not already filled.
A preferable solution might be to pass the complete list of fields (e.g., HazelcastObjectType.fields and QueryDataType.objectFields) to the constructor. I examined all code parts where fields are added, and it seems that they can be gathered before creating the object.
These alternatives aim to make the code more readable, less error-prone, and logically sound.

Reply:
1. You are right. If a list cannot be traversed, then it is a list builder. It doesn't make sense for it to implement `List` interface. Imitating an interface by providing empty/throwing implementations is a strong indicator of poor design as I also mentioned [here][1].
2. The creator of the type can build the field list and pass it to the `addAll` method. However, in this way, we would have to keep track of field lists. For example, `typeMap` would be `Map<String, Tuple2<QueryDataType, List<QueryDataTypeField>>>`. By following OOP principles, encapsulation particularly, we simplify the logic.
3. This is impossible because of cyclic types. If `TypeA` contains `TypeB`, and `TypeB` contains `TypeA`, which one will you create first?

I will follow your first suggestion and replace `NonTraversableList` with `ImmutableListBuilder`, which doesn't implement `List` interface. The type classes will contain two fields: one for the builder and one for the finalized field list. The finalized field list will be null initially, and as a consequence, it won't be traversable until finalization. The builder will be set to null at finalization, and as a consequence, the caller won't be able to add new fields after finalization.

[1]: https://github.com/hazelcast/hazelcast/pull/26056#discussion_r1416323180